DOC: Improve groupby().ngroup() explanation for missing groups

NickCrews opened the pull request:
The exact prose and examples are open to improvements if you have ideas.

I figure this didn't require a separate issue, please let me know if I'm missing some step in the process.

NickCrews committed:
DOC: Improve groupby().ngroup() explanation for missing groups

mroeschke commented on pandas/core/groupby/groupby.py:
```diff
@@ -3228,15 +3231,17 @@ def ngroup(self, ascending: bool = True):
         Examples
         --------
-        >>> df = pd.DataFrame({"A": list("aaabba")})
+        >>> df = pd.DataFrame()
```
Nit: Could you combine the DataFrame construction in 1 line (construct the DataFrame from dictionary?)

datapythonista commented on pandas/core/groupby/groupby.py:
```diff
@@ -3212,6 +3212,9 @@ def ngroup(self, ascending: bool = True):
         would be seen when iterating over the groupby object, not the
         order they are first observed.
 
+        If a group would be excluded (due to null keys) then that
+        group is labeled as np.nan. See examples below.
```
I think it could be simpler to understand if we just simply say something like `Groups where the key is missing are skipped from the count, and their value will be 'NaN'`.

rhshadrach replied:
I'd prefer NA as opposed to "missing". There are various reasons a value can be NA, missing is but one of them. Also recommend using NA over NaN or null as it matches `dropna` and `pd.NA`.

datapythonista replied:
Fine by me. But in this case if we say `NA` it may sound like `pandas.NA`, and if the value is `None` or`NaN` it will also be skipped. So, missing seemed more appropriate to me.

datapythonista replied:
Maybe `not available` instead of `NA` is more generic than missing, and doesn't sound like we are talking about `pandas.NA`?

rhshadrach replied:
Sorry - that was poorly written. My main reason is to connect it with the `dropna` argument. I threw in `pd.NA` as an indicator that we use NA with somewhat good consistency. E.g. dropna, pd.NA, isna, use_inf_as_na, fillna, na_filter. I think we should maintain this.

NickCrews replied:
I adjusted to "Groups with missing keys (where `pd.isna()` is True) will be labeled with `NaN` and will be skipped from the count." Definitely this is the right direction, I'm still just not sure about the "missing" language

I agree with

> But in this case if we say NA it may sound like pandas.NA, and if the value is None orNaN it will also be skipped.

so I avoided saying simply "groups with NA keys...". I think my phrasing is maybe a little longer, but extremely precise and fairly easy to understand. Another option I would also be satisfied with is "Groups with NA keys (where `pd.isna()` is True)"

rhshadrach replied:
Thanks - this looks great to me.

datapythonista commented on pandas/core/groupby/groupby.py:
```diff
@@ -3228,15 +3231,17 @@ def ngroup(self, ascending: bool = True):
         Examples
         --------
-        >>> df = pd.DataFrame({"A": list("aaabba")})
+        >>> df = pd.DataFrame()
```
Agree with Matt, also I think the examples can be simplified. I think having just a single column with the `None` value is enough, no need to show without the `None` first, and with it later.

Also, unrelated to your changes, but I'd remove `df.groupby(["A", [1,1,2,3,2,1]]).ngroup()`. This is a very specific case, not directly related to `ngroup`, and grouping with provided values is not even shown in the examples of `groupby`, so it makes less sense to show it here. I think it mostly adds confusion to this docstring.

And also, I think it's better for users if the examples are a bit more meaningful than random letters. I think using something like `df = pd.DataFrame({'color': ['red', 'red', 'green', None, 'red', 'green']})` (or with animals, jobs, whatever... would make the example a bit easier to understand. Just as an idea, if you want to keep this PR only for the example you're trying to illustrate, that's surely fine.

NickCrews replied:
I agree with all of these suggestions. fixup PR incoming.

datapythonista reviewed (commented):
Thanks @NickCrews, nice improvement. Added couple of comments. Sorry for the late review, I hope you still have time to address comments, would be great to see this merged.

NickCrews committed:
DOC: fixup PR suggestions

Per comments at
https://github.com/pandas-dev/pandas/pull/50049

NickCrews commented on pandas/core/groupby/groupby.py:
```diff
@@ -3228,39 +3231,31 @@ def ngroup(self, ascending: bool = True):
+        4    1.0
+        5    0.0
+        dtype: float64
+        >>> df.groupby("color", dropna=False).ngroup()
```
I dropped the `ascending` example, I figured that was obvious enough. Think I should keep it in?

rhshadrach replied:
I think we typically demonstrate all the features of a method, regardless of how obvious they might seem. I think it should be retained.

NickCrews replied:
I restored
```
        >>> df.groupby("color", dropna=False).ngroup(ascending=False)
        0    1
        1    0
        2    1
        3    2
        4    2
        5    1
        dtype: int64
```

I chose to use `dropna=False` because I wanted to show
that NA keys are placed BEFORE other keys.

I figured the `dropna=True` example was obvious enough from this and
I didn't need that one as well.

NickCrews committed:
DOC: fixup: update order of labels

Now I guess since the groups
are lexicographically sorted,
and we are using "red" and "blue" instead of "a" and "b",
the ngroup labels have swapped order.

I think therefore that this should be deterministic
and not flaky.

NickCrews commented:
That failing test looks unrelated to this change.

datapythonista reviewed (approved):
lgtm, not strong opinion about the ascending example, from my side is fine both with and without it. Thanks @NickCrews

NickCrews committed:
DOC: fixup: restore ascending=false example in docstring

I chose to use `dropna=False` because I wanted to show
that NA keys are placed
BEFORE other keys. I figured
the `dropna=True` example was obvious enough from this and
I didn't need that one as well, otherwise I thought things
got very verbose.

mroeschke reviewed (approved):
LGTM merge when ready @rhshadrach

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @NickCrews
